Step2: QueryBuilder DDL

moonimooni opened the pull request:
안녕하세요, 

이런 저런 구조적인 고민을 많이 하게 되네요
커밋 보시면 요구사항1 까지는 생각의 흐름대로 쭈욱 작성하다가, 
요구사항2부터 복잡해지는 거 같아 맵핑모델을 만들어야겠다 싶어서 EntityColumn, EntityTable 등등... 을 추가했습니다.

어플리케이션 실행 결과물 캡쳐 첨부합니다.
<img width="765" alt="Screenshot 2024-10-12 at 1 18 32 AM" src="https://github.com/user-attachments/assets/17b48195-2580-4968-af77-055b2da03ddc">

추가적인 코멘트는 코드에 직접 달겠습니다!

moonimooni committed:
doc: step 2 요구사항 정리

moonimooni committed:
test: 요구사항 1 테스트 작성

moonimooni committed:
feat: 요구사항 1 기능 추가

moonimooni committed:
test: 요구사항 2 테스트 작성

moonimooni committed:
feat: 요구사항 2 기능 작성

- `@Entity` 어노테이션을 지닌 클래스의 정보를 맵핑하는 객체 구조 구성-

moonimooni committed:
test: 요구사항 2 유닛 테스트 추가

moonimooni committed:
test: 요구사항 3 기능 작성

moonimooni committed:
test: 요구사항 4 기능 작성

moonimooni committed:
refactor: 컬럼 length 설정 가능하도록 함

moonimooni commented on src/main/java/persistence/model/meta/DataType.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Arrays;
+
+public enum DataType {
+    VARCHAR(String.class, "varchar(%d)", 255),
```
enum은 확장이 어렵다고 생각이 들긴 합니다.
(좀 멀리 생각하면 varchar2 이런 키워드 쓰는 디비들도 있으니..)

그러다 [피드백 페이지](https://edu.nextstep.camp/s/BkhC9zSg/ls/SHL2FcGN)랑 [hibernate 코드](https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java#L408) 보니 각 데이터베이스 dialect별로 typeRegistry에 타입을 추가해주는 거 같아서, 비슷하게 시도해 보다가 현재 규모에 안맞게 복잡해지는 거 같아서 일단 이대로 진행했습니다.

leeyohan93 replied:
저도 실무에서는 규모에 맞는 개발을 굉장히 중요시 하는데요 :)
JP 과정동안에는 교육이기 때문에 라이브러리를 만들어본다는 마음으로 조금 더 도전해보셔도 좋을 것 같습니다! 😄 

희원님은 구현 역량도 충분히 있으신 것 같아 더 많은 요구사항을 드릴 수도 있을 것 같아요 :)
추가적으로 이야기드리자면 DB 마다 타입이 다르게 구현되는데 이러한 요구사항도 받아들일 수 있도록 구조를 개선해보는건 어떨까요~?
```
정수 타입
- MySQL: INT
- PostgreSQL: INTEGER
```

moonimooni replied:
넵 언어별로 다르게 쓰일 수 있도록 dialect가 데이터타입을 관리하도록 수정해보았습니다!

moonimooni commented on src/main/java/persistence/sql/dialect/H2Dialect.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public String getAutoGeneratedIdentityPhrase() {
+        return "generated by default as IDENTITY";
```
검색해보니 H2는 이렇게 쓰는게 기본 문법인 것 같아서 시도해보았습니다 🙃

moonimooni commented on src/main/java/persistence/model/meta/Constraint.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public abstract class Constraint {
```
__시간 나면__ unique도 구현해볼 수 있겠다 생각하여 abstract class로 구현했습니다.

moonimooni commented on src/main/java/persistence/model/EntityColumn.java:
```diff
@@ -0,0 +1,86 @@
+public class EntityColumn {
+    private final String name;
+
+    private final Field field;
```
여기 Field 필드를 들고 있는게 맞는지 고민이 되네요.

데이터베이스 모델 관점에서 생각하면 빠지는 게 맞는데,
매퍼 역할이라 생각하면 틀린 것도 아닌 거 같고

isNullable(), isPrimary() 이런 메소드에서 쓰이고 있긴 한데,
차라리 이런 Boolean 값을 멤버변수로 선언하고 생성자에서 할당해주는 게 더 나을지... 🤔

leeyohan93 replied:
개인적인 의견으로는 아래와 같은 이유로 희원님이 이야기해주신 `Boolean 값을 멤버변수로 선언하고 생성자에서 할당` 방법이 더 괜찮은 방법이라고 생각합니다 :)

1. 리플렉션에 의존적이지 않은 코드 구현 -> POJO 기반 구현
2. 객체 초기화 시에 처음에만 리플렉션 코드를 호출 -> 성능적인 부분

moonimooni replied:
감사합니다 수정했습니다~

leeyohan93 commented on src/main/java/persistence/model/EntityTable.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.List;
+
+public class EntityTable {
```
`복잡해지는 거 같아 맵핑모델을 만들어야겠다 싶어서 EntityColumn, EntityTable 등등... 을 추가했습니다.`

굉장히 좋은 접근인 것 같습니다. 책임에 맞게 적절하게 클래스를 나누어주신 것 같아요! 👍 
다만 EntityTable과 EntityColumn 등에 대한 테스트가 없는 것이 아쉽습니다 😢 
앞으로 과정을 진행하며 테스트가 큰 도움이 되니 함께 작성해주시면 좋을 것 같습니다 :)

moonimooni replied:
유닛테스트 더 추가했습니다~!

leeyohan93 commented on src/main/java/persistence/model/util/ReflectionUtil.java:
```diff
@@ -0,0 +1,21 @@
+import java.lang.reflect.Field;
+import java.util.Optional;
+
+public class ReflectionUtil {
+    public static <T extends Annotation> Optional<T> getAnnotationIfPresent(Field field, Class<T> annotationClass) {
+        if (field.isAnnotationPresent(annotationClass)) {
+            return Optional.ofNullable(field.getAnnotation(annotationClass));
+        }
+        return Optional.empty();
+    }
```
리플렉션과 Optional을 잘 사용해서 사용성 좋은 유틸을 만들어주셨네요! 💯

leeyohan93 commented on src/main/java/persistence/model/EntityTable.java:
```diff
@@ -0,0 +1,48 @@
+
+    private final EntityTableColumns columns;
+
+    private final Constraint primaryConstraint;
```
테이블이 기본 키를 가져야하는 제약조건을 표현하기 위해 객체 값을 추가해주셨네요! 👍 
다만 Constraint는 현재 단순히 EntityColumn을 목록으로 가지는 DTO와 같은 역할을 하고 있는 것 같아요 🤔 

클래스를 작게 분리하는 것은 정말 중요하지만, 적절한 책임에의해 만들어져야하며 항상 오버엔지니어링은 경계해야합니다 :)
적절한 책임을 가지고 있는지 점검해보시면 좋을 것 같습니다 :)

moonimooni replied:
넵 듣고보니 그렇네요 ㅎㅎ
일단 Constraint 관련 코드들 제거했습니다. 나중에 필요해지면 다시 추가하든 해야겠네요.
대신 EntityTableColumns로부터 필요한 컬럼 추리는 것으로 정리했어요.

leeyohan93 commented on src/main/java/persistence/sql/ddl/DdlQueryBuilder.java:
```diff
@@ -0,0 +1,71 @@
+                .collect(Collectors.joining(", "));
+    }
+
+    private String getColumnDdl(EntityColumn column) {
+        String nameQuoted = dialect.getIdentifierQuoted(column.getName());
+        String dataType = column.getDataType().getFullName(column.getLength());
+        String nullPhrase = dialect.getNullPhrase(column.isNullable());
+
+        if (column.isAutoGeneratedIdentity()) {
+            return getAutoGeneratedIdentityColumnDdl(nameQuoted, dataType, nullPhrase);
+        }
+        return nameQuoted + " " + dataType + " " + nullPhrase;
+    }
```
QueryBuilder 에서는 컬럼의 정보들을 통해 SQL을 구성하는 책임을 적절하게 분리해주신 것 같아요~ 💯 💯

leeyohan93 commented on src/main/java/persistence/sql/dialect/H2Dialect.java:
```diff
@@ -0,0 +1,54 @@
+        return "generated by default as IDENTITY";
+    };
+
+    @Override
+    public String getCreateTablePhrase() {
+        return "CREATE TABLE";
+    }
```
CREATE, SELECT, UPDATE, DROP 등은 데이터베이스마다 다르지 않은 SQL 표준 문법으로 다루어도 괜찮지 않을까요~?
물론 극단적으로 다른 데이터베이스가 있다고 가정하고 지금 구조를 유지하셔도 괜찮습니다 :)
희원님 의견이 궁금했어요! :)

moonimooni replied:
예를 들어 `DROP TABLE IF EXISTS`는 아직 지원안되는 디비도 있고, 되는 디비도 특정 버전은 지원 안되는걸로 알고 있어서 Dialect로 넣었어요. 안그래도 Dialect가 인터페이스면 많은 구현체들에서 코드가 중복될까 고민되었는데 abstract class로 바꾸면서 Dialect 내에서 표준 문법 기준을 세울 수 있게 됐습니다.

leeyohan93 reviewed (changes requested):
안녕하세요 희원님! 😄 
쿼리빌더 요구사항에 맞게 잘 구현해주셨습니다! 👍 
많은 고민을 해보신 흔적이 느껴져서 저도 같이 열심히 고민해보았던 것 같아요 :)
다음 단계에 넘어가기 전에 개선해보면 좋을 부분에 코멘트를 남겨두었습니다 😄 
주말 날씨가 좋네요~ 좋은 주말 보내시면서 미션 화이팅입니다 🔥

moonimooni committed:
refactor: 피드백 반영

- Constraint 클래스 제거

- EntityTable, EntityTableColumns 유닛테스트 추가

- Dialect에 DataTypeRegistry 추가

moonimooni committed:
refactor: EntityFactory 추가

- EntityTable에서 각 컬럼 생성 로직을 담당해야 하는 책임을 덜어냄

moonimooni committed:
refactor: .gitignore에 bin 추가

moonimooni commented on src/main/java/persistence/model/EntityFactory.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class EntityFactory {
```
컬럼이 dataType을 가지고 있다는 관점으로 코드를 구현했고, 
dataType을 알아내려면 dialect가 필요하기 때문에 (dataTypeRegistry를 추가함으로 인해) 컬럼 객체 생성할 때 dialect가 필요하게 되었습니다.

그렇게 되면 EntityTable.build할때부터 dialect 인자를 쭈욱 전달해줘야 했는데 EntityTable이나 EntityTableColumns는 dialect가 필요 없었어요.
그래서 아예 별도의 생성 팩토리를 분리해보았습니다.

leeyohan93 replied:
Factory를 통해 필요하지 않은 의존성을 제거해주신 접근이 좋네요!! 👍 
다만 EntityColumn이 dataType을 상태로 가지고 있어야 하는가? 에 대해서는 고민이 필요한 부분인 것 같습니다.

데이터베이스 또는 SQL과 관련된 변경에서 엔티티는 변경되지 않는 구조를 만들어보는건 어떨까요? :)

leeyohan93 commented on src/main/java/persistence/sql/dialect/Dialect.java:
```diff
@@ -0,0 +1,95 @@
+
+import static java.sql.Types.*;
+
+public abstract class Dialect {
```
상속을 사용하게되면 자식 클래스에서 부모 클래스의 기능을 알아야한다는 단점이 있는데요!
상속 보다는 합성을 적절히 사용해보면 어떨까요! 😄 

좋은 포스팅이 있어 함께 공유드립니다 :)
https://mangkyu.tistory.com/199

leeyohan93 replied:
희원님~ 해당 코멘트는 제가 잘못드린 가이드인 것 같습니다!
지금과 같이 추상클래스를 통한 상속이 더욱 잘어울리는 케이스 같네요 😂

leeyohan93 reviewed (approved):
안녕하세요 희원님!
피드백 잘 적용해서 개선해주셨네요! 👍 
다음 단계 진행하면서 함께 고민해보면 좋을 코멘트를 남겨두었습니다.
궁금한 점이 있으시다면 편하게 DM 남겨주세요!
이번 한주도 화이팅 입니다! 😄
